cli/server_test.go

```diff
@@ -0,0 +1,129 @@
+		cfg := *ptr
+		chainPath := filepath.Join(t.TempDir(), "neogotestchain")
+		cfg.ApplicationConfiguration.DBConfiguration.Type = "leveldb"
+		cfg.ApplicationConfiguration.DBConfiguration.LevelDBOptions.DataDirectoryPath = chainPath
+		f(&cfg)
+		out, err := yaml.Marshal(cfg)
+		require.NoError(t, err)
+
+		cfgPath := filepath.Join(tmpDir, "protocol.unit_testnet.yml")
+		require.NoError(t, ioutil.WriteFile(cfgPath, out, os.ModePerm))
+		t.Cleanup(func() {
+			require.NoError(t, os.Remove(cfgPath))
+		})
+		return cfgPath
+	}
+
+	baseCmd := []string{"neo-go", "node", "--unittest", "--config-path", tmpDir}
+	e := newExecutor(t, false)
+
+	t.Run("invalid config path", func(t *testing.T) {
+		e.RunWithError(t, baseCmd...)
+	})
+	t.Run("bad logger config", func(t *testing.T) {
+		badConfigDir := t.TempDir()
+		logfile := filepath.Join(badConfigDir, "logdir")
+		require.NoError(t, os.WriteFile(logfile, []byte{1, 2, 3}, os.ModePerm))
+		saveCfg(t, func(cfg *config.Config) {
+			cfg.ApplicationConfiguration.LogPath = filepath.Join(logfile, "file.log")
+		})
+		e.RunWithError(t, baseCmd...)
+	})
+	t.Run("invalid storage", func(t *testing.T) {
+		saveCfg(t, func(cfg *config.Config) {
+			cfg.ApplicationConfiguration.DBConfiguration.Type = ""
+		})
+		e.RunWithError(t, baseCmd...)
+	})
+	t.Run("stateroot service is on && StateRootInHeader=true", func(t *testing.T) {
+		saveCfg(t, func(cfg *config.Config) {
+			cfg.ApplicationConfiguration.StateRoot.Enabled = true
+			cfg.ProtocolConfiguration.StateRootInHeader = true
+		})
+		e.RunWithError(t, baseCmd...)
+	})
+	t.Run("invalid Oracle config", func(t *testing.T) {
+		saveCfg(t, func(cfg *config.Config) {
+			cfg.ApplicationConfiguration.Oracle.Enabled = true
+			cfg.ApplicationConfiguration.Oracle.UnlockWallet.Path = "bad_orc_wallet.json"
+		})
+		e.RunWithError(t, baseCmd...)
+	})
+	t.Run("invalid consensus config", func(t *testing.T) {
+		saveCfg(t, func(cfg *config.Config) {
+			cfg.ApplicationConfiguration.UnlockWallet.Path = "bad_consensus_wallet.json"
+		})
+		e.RunWithError(t, baseCmd...)
+	})
+	t.Run("invalid Notary config", func(t *testing.T) {
+		t.Run("malformed config", func(t *testing.T) {
+			saveCfg(t, func(cfg *config.Config) {
+				cfg.ProtocolConfiguration.P2PSigExtensions = false
+				cfg.ApplicationConfiguration.P2PNotary.Enabled = true
+			})
+			e.RunWithError(t, baseCmd...)
+		})
+		t.Run("invalid wallet", func(t *testing.T) {
+			saveCfg(t, func(cfg *config.Config) {
+				cfg.ProtocolConfiguration.P2PSigExtensions = true
+				cfg.ApplicationConfiguration.P2PNotary.Enabled = true
+				cfg.ApplicationConfiguration.P2PNotary.UnlockWallet.Path = "bad_notary_wallet.json"
+			})
+			e.RunWithError(t, baseCmd...)
+		})
+	})
+	t.Run("good", func(t *testing.T) {
+		var (
+			rpcAddress string
+			rpcPort    uint16
+		)
+		saveCfg(t, func(cfg *config.Config) {
+			// firstly, get free port for RPC server
+			addr, err := net.ResolveTCPAddr("tcp", "localhost:0")
```

Comment:
Not reliable, it can be occupied by another process between `l.Close` and `e.Run`. IIRC we have logic in RPC itself to handle 0 properly already (choosing the port dynamically).